### ISSUE TITLE
Let datasource plugins manage folders and dashboards

### ISSUE THREAD

**Issue by patrickjuchli** [grafana/grafana#19435]
**What would you like to be added**:

A datasource (or app?) plugin can add, edit and remove folders and dashboards. This could also be limited by scope, so that the plugin can't do anything to folders/dashboards that it didn't also create.

**Why is this needed**:

I'm using a custom datasource plugin to access a custom database. Access is granted using credentials provided in the plugin configuration.

Let's call the things that I show metrics for "systems". Access credentials entered above are associated with a user account that is associated with a specific set of systems that I now have access to. 

I want to show dashboards for these systems. Typically, I'd use dashboard variables to switch between systems and placeholders in my dashboard(s). Unfortunately, my systems are very heterogenous, they don't all have the same structure and metrics. Dashboards need to be somewhat tailored to the specific systems. 

So what I need is probably more something like folders for each system and in these folders multiple dashboards. If I were to write my own Frontend, this would be quite easy. My data source knows how this should be done because it also knows about the structure of each system. Defining dashboards based on that information can be automated by custom code. Dashboards would be generated on-the-fly.

But how do I do this with Grafana? I can't define folders/dashboards with my datasource plugin.

I've looked at dashboard provisioning. I could use a dashboard provider per system/folder that links to system-specific dashboard files. But where exactly are these things setup and updated? Also, I have to call `/api/admin/provisioning/dashboards/reload` if there is some change in the set of system/folders. My use-case seems to be a bit too dynamic for provisioning.

I've looked at the Grafana API but still, I need to "prepare" things. I need to write another piece of code that syncs the state of folders/dashboards on Grafana with the state of the systems in my database. This is getting complex and error-prone. Especially if I also have different users who have access to different systems where I'd probably have to introduce organizations.

All of this is absolutely no problem with a custom Frontend that dynamically populates lists of available systems and dashboards given a user and the data accessible. But with Grafana, this seems to be very hard.

I hope I've explained my use-case well enough. Please do complain if not because this problem might be the reason for us (my company) not to use Grafana. This is not a complaint, I absolutely understand if you want Grafana to focus on other use-cases

**Comment by patrickjuchli** [comment]
Another way to describe this issue:

I think Grafana's dashboards are things with state – probably to allow annotations and alarms associated/edited with them. The consequence of what I'm asking for is probably stateless dashboards, that are created and removed on-the-fly. This might be running counter to the core design of Grafana.

**Comment by ryantxu** [comment]
I don't totally understand what you are trying to do... and I'm not sure it is not already possible :)

Have you seen https://grafana.com/docs/reference/scripting/

also you can create/edit etc dashboads with the API https://grafana.com/docs/http_api/dashboard/

**Comment by patrickjuchli** [comment]
Sorry for the confusion, let me try again:

I have a custom backend. It holds my metrics for all systems but also offers user management. Users have access to specific sets of systems. "System" is a thing that exposes some metrics.

Users should be able to use Grafana to look at metrics of systems they have access to.

I've written a datasource plugin connecting to this backend. I've been using dashboard variables to switch between systems but learned that systems differ quite a bit and I need to use folders for each system populated with custom sets of dashboards. These custom sets can still be created automatically, though.

How do I do that with Grafana?

There are scripted dashboards, provisioning and the API.

Scripted dashboards don't seem to be used that much and Grafana might deprecate them [according to this issue](https://github.com/grafana/grafana/issues/14598#issuecomment-448640803). Is this true?

There's a lot of activity around provisioning. But to have updates to dashboard providers considered, I need to use `/api/admin/provisioning/dashboards/reload`. Using the API directly seems to be a better fit.

The API absolutely lets me do what I want. But it ends up being very complicated. Next to the datasource plugin, another mechanism has to prepare all possible folders/dashboards using the API beforehand and needs to keep all this in sync for multiple users having access to different sets of systems.

As a contrast, here is what a custom Frontend would do:

Using my backend API, display a list of systems the logged-in user has access to. (In Grafana, that would be "folders"). For each system, dynamically set the available overviews (In Grafana "dashboards" in these folders). It doesn't get much simpler than that.

The difference with Grafana is, that folders/dashboards have all to be defined "statically" in the database using the API before any user is accessing Grafana. Or is there another way?

**Comment by patrickjuchli** [comment]
An answer is not required anymore, cheers stale bot.
